src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,253 @@
+import {Logger, Messages, SfdxError} from '@salesforce/core';
+import {xml2js} from 'xml-js';
+import {Controller} from '../../Controller';
+import {Catalog, Rule, RuleGroup, RuleResult, RuleTarget, RuleViolation, TargetPattern} from '../../types';
+import {AbstractRuleEngine} from '../services/RuleEngine';
+import {Config} from '../util/Config';
+import {ENGINE, Severity} from '../../Constants';
+import * as engineUtils from '../util/CommonEngineUtils';
+import CpdWrapper from './CpdWrapper';
+import {FILE_EXTS_TO_LANGUAGE} from '../../Constants';
+import {uxEvents} from '../ScannerEvents';
+import crypto = require('crypto');
+
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/sfdx-scanner', 'CpdEngine');
+
+export class CpdEngine extends AbstractRuleEngine {
+
+	public readonly ENGINE_ENUM: ENGINE = ENGINE.CPD;
+	public readonly ENGINE_NAME: string = ENGINE.CPD.valueOf();
+
+	private minimumTokens: number;
+	private logger: Logger;
+	private config: Config;
+	private initialized: boolean;
+	private cpdCatalog: Catalog;
+	
+
+
+	public getName(): string {
+		return this.ENGINE_NAME;
+	}
+
+	public getTargetPatterns(): Promise<TargetPattern[]> {
+		return this.config.getTargetPatterns(this.ENGINE_ENUM);
+	}
+
+	public getCatalog(): Promise<Catalog>{
+		return Promise.resolve(this.cpdCatalog);
+	}
+
+	/* eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars */
+	public shouldEngineRun(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string, string>): boolean {
+		return true;
+	}
+	
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	public async run(ruleGroups: RuleGroup[], rules: Rule[], targets: RuleTarget[], engineOptions: Map<string, string>): Promise<RuleResult[]> {
+
+		const languageToPaths = this.sortPaths(targets);
+		const results = [];
+
+		if (languageToPaths.size === 0) {
+			this.logger.trace(`No matching cpd target files found. Nothing to execute.`);
+			return [];
+		}
+		
+		const ps: Promise<string>[] = [];
+		// cpd only allows one language to be speificed at a time, so we must run it for each
+		for (const language of Array.from(languageToPaths.keys())) {
+			try {
+				ps.push(this.runLanguage(language, languageToPaths.get(language)));
+			} catch (e) {
+				this.logger.trace(`Cpd (${language}) evaluation failed: ` + (e.message || e));
```

Comment:
I think the `catch` may not serve the intended purpose since the try only guards the `push` method and not the actual `run`. This might have to be within the `runLanguage` method.